cmd/ctr/commands/images/images.go

```diff
@@ -355,3 +357,77 @@ var removeCommand = cli.Command{
 		return exitErr
 	},
 }
+
+var pruneCommand = cli.Command{
+	Name:  "prune",
+	Usage: "remove unused images",
+	Flags: []cli.Flag{
+		cli.BoolFlag{
+			Name:  "all", // TODO: add more filters
+			Usage: "remove all unused images, not just dangling ones (if all is not specified no images will be pruned)",
+		},
+	},
+	// adapted from `nerdctl`:
+	// https://github.com/containerd/nerdctl/blob/272dc9c29fc1434839d3ec63194d7efa24d7c0ef/cmd/nerdctl/image_prune.go#L86
+	Action: func(context *cli.Context) error {
+		client, ctx, cancel, err := commands.NewClient(context)
+		if err != nil {
+			return err
+		}
+		defer cancel()
+
+		all := context.Bool("all")
+		if !all {
+			log.G(ctx).Warn("No images pruned. `image prune` requires --all to be specified.")
+			// NOP
+			return nil
+		}
+
+		var (
+			imageStore     = client.ImageService()
+			contentStore   = client.ContentStore()
+			containerStore = client.ContainerService()
+		)
+		imageList, err := imageStore.List(ctx)
+		if err != nil {
+			return err
+		}
+		containerList, err := containerStore.List(ctx)
+		if err != nil {
+			return err
+		}
+		usedImages := make(map[string]struct{})
+		for _, container := range containerList {
+			usedImages[container.Image] = struct{}{}
+		}
+
+		delOpts := []images.DeleteOpt{images.SynchronousDelete()}
+		removedImages := make(map[string][]digest.Digest)
+		for _, image := range imageList {
+			if _, ok := usedImages[image.Name]; ok {
+				continue
+			}
+
+			digests, err := image.RootFS(ctx, contentStore, platforms.DefaultStrict())
+			if err != nil {
+				log.G(ctx).WithError(err).Warnf("failed to enumerate rootfs of image %s", image.Name)
+			}
+			if err := imageStore.Delete(ctx, image.Name, delOpts...); err != nil {
+				log.G(ctx).WithError(err).Warnf("failed to delete image %s", image.Name)
+				continue
+			}
+			removedImages[image.Name] = digests
+		}
+
+		if len(removedImages) > 0 {
+			log.G(ctx).Infoln("deleted images:")
+			for image, digests := range removedImages {
+				log.G(ctx).Infof("untagged: %s\n", image)
+				for _, digest := range digests {
+					log.G(ctx).Infof("deleted: %s\n", digest)
```

Comment:
I don't think this can be assumed and not something that is surfaced to the client. With `ctr` I think its best to error on accuracy or skip the functionality. The containerd logs would accurately show what got removed during GC. I don't think it is worth it here to check each layer to see if it still exists.

Reply:
Thanks for review @dmcgowan. I removed the layer check/log here.